task_geo/data_sources/covid/fr_covidata/fr_covidata.py

```diff
@@ -0,0 +1,257 @@
+                dataset.loc[i + 1, 'reanimation'] = 'inv'
+
+        if dataset.loc[i, 'maille_code'] == dataset.loc[
+            i + 1, 'maille_code'] and dataset.loc[i, 'date'] == dataset.loc[
+                i + 1, 'date'] and dataset.loc[i, 'hospitalises'] != 'inv':
+            if pd.isnull(
+                dataset.loc[i, 'hospitalises']) is True and pd.isnull(
+                    dataset.loc[i + 1, 'hospitalises']) is False:
+                dataset.loc[i, 'hospitalises'] = dataset.loc[
+                    i + 1, 'hospitalises']
+                dataset.loc[i + 1, 'hospitalises'] = 'inv'
+            elif pd.isnull(
+                    dataset.loc[i, 'hospitalises']) is False and pd.isnull(
+                    dataset.loc[i + 1, 'hospitalises']) is True:
+                dataset.loc[i + 1, 'hospitalises'] = 'inv'
+            elif pd.isnull(
+                dataset.loc[i, 'hospitalises']) is True and pd.isnull(
+                    dataset.loc[i + 1, 'hospitalises']) is True:
+                dataset.loc[i + 1, 'hospitalises'] = 'inv'
+            elif dataset.loc[i, 'hospitalises'] == dataset.loc[
+                    i + 1, 'hospitalises']:
+                dataset.loc[i + 1, 'hospitalises'] = 'inv'
+            elif dataset.loc[i, 'hospitalises'] > dataset.loc[
+                    i + 1, 'hospitalises']:
+                dataset.loc[i + 1, 'hospitalises'] = 'inv'
+            elif dataset.loc[i, 'hospitalises'] < dataset.loc[
+                    i + 1, 'hospitalises']:
+                dataset.loc[i, 'hospitalises'] = dataset.loc[
+                    i + 1, 'hospitalises']
+                dataset.loc[i + 1, 'hospitalises'] = 'inv'
+
+        if dataset.loc[i, 'maille_code'] == dataset.loc[
+            i + 1, 'maille_code'] and dataset.loc[i, 'date'] == dataset.loc[
+                i + 1, 'date'] and dataset.loc[i, 'gueris'] != 'inv':
+            if pd.isnull(dataset.loc[i, 'gueris']) is True and pd.isnull(
+                    dataset.loc[i + 1, 'gueris']) is False:
+                dataset.loc[i, 'gueris'] = dataset.loc[i + 1, 'gueris']
+                dataset.loc[i + 1, 'gueris'] = 'inv'
+            elif pd.isnull(dataset.loc[i, 'gueris']) is False and pd.isnull(
+                    dataset.loc[i + 1, 'gueris']) is True:
+                dataset.loc[i + 1, 'gueris'] = 'inv'
+            elif pd.isnull(dataset.loc[i, 'gueris']) is True and pd.isnull(
+                    dataset.loc[i + 1, 'gueris']) is True:
+                dataset.loc[i + 1, 'gueris'] = 'inv'
+            elif dataset.loc[i, 'gueris'] == dataset.loc[i + 1, 'gueris']:
+                dataset.loc[i + 1, 'gueris'] = 'inv'
+            elif dataset.loc[i, 'gueris'] > dataset.loc[i + 1, 'gueris']:
+                dataset.loc[i + 1, 'gueris'] = 'inv'
+            elif dataset.loc[i, 'gueris'] < dataset.loc[i + 1, 'gueris']:
+                dataset.loc[i, 'gueris'] = dataset.loc[i + 1, 'gueris']
+                dataset.loc[i + 1, 'gueris'] = 'inv'
```

Comment:
I'm not sure about what you want to accomplish here, but I'm pretty sure `pandas` provides with cleaner and simpler methods to do so.

I would like top help you, so please can you:
1. Show me how the data is before entering the loop, something like the output of `dataset.head(3).T`
2. Explain me in a detailed way what you want to accomplish.
3. Show me the data after exiting the for loop with a `dataset.head(3).T`

Reply:
Thank you for your valuable inputs.
The dataset contains raw data for the same region & date, from different sources.
The issue here is that none of those sources info can be omitted, and so the optimal and only way is to aggregate that data.
The other issue is that for numerous cases, same region and date, the numbers are different depending on the source; therefore an assumption had to be made.
The assumption as explained is that, for each column, the higher value of those different numbers is taken into consideration.
Therefore, the above script iterates through the rows and updates them.
In case there are two rows with the same date and region, it checks if any of the values is null:
- if both values are 'null', then the first-row value stays null, and the second row takes the value 'inv'.
- if one of the values is not 'null', it is assigned to the first row and the second row takes the value 'inv'.
- if both are not 'null', the first-row value is the highest of both values, and the second row takes the value 'inv'.

These are done for all the numerical columns. And in all those cases, the sources info (name, url, type) are added together in the first row - that is why I had to change their type to string.
Eventually, all those redundant rows containing the value 'inv' are dropped.

Input example:
![example](https://user-images.githubusercontent.com/22396992/78678969-4e578a00-78f2-11ea-8c10-ec39f6d8b333.PNG)

Output:
![example2](https://user-images.githubusercontent.com/22396992/78679112-7c3cce80-78f2-11ea-9222-dc4ba3f97b9a.PNG)

All the other changes have been implemented.

Reply:
> The issue here is that none of those sources info can be omitted, and so the optimal and only way is to aggregate that data.

I think that this is not the correct approach to take, as you are hand-picking data.

What I would do instead is select, for each region and column, which source you consider more reliable ( usually the one that has more data) and grab ALL values from that source, and don't fill missing values.

>  The other issue is that for numerous cases, same region and date, the numbers are different depending on the source; therefore an assumption had to be made.
The assumption as explained is that, for each column, the higher value of those different numbers is taken into consideration.

I think that any assumption should be done looking at how much data a source brings, not in the values, let me think a little about it and will come with a proposal.

Reply:
Hi @oussamanaji, I haven't forgotten about your review, I've been discussing it with @DanielRobertNicoud, will get you an update soon, on the meanwhile, please address all other comments on the review.

Thanks.

Reply:
Hello @ManuelAlvarezC, thank you for the update.
I have already fixed all the other issues. Should I update the file and add the others meanwhile?
Thank you

Reply:
Yes, please

Reply:
Hi @oussamanaji,
After consulting it with other people we have decided that your approach, altough maybe a little controversial, will be the one we will be using for now.

What will be changing is the way is implemented, instead of all this lines of code, you can replace all that with the following:
```python
aggregations = {
    'cas_confirmes': np.max,
    'cas_ehpad': np.max,
    'cas_confirmes_ehpad': np.max,
    'cas_possibles_ehpad': np.max,
    'deces': np.max,
    'deces_ehpad': np.max,
    'reanimation': np.max,
    'hospitalises': np.max,
    'gueris': np.max,
    'source_nom': ','.join,
    'source_url': ','.join,
    'source_archive': ','.join,
    'source_type': ','.join
}

dataset = dataset.groupby(['date', 'maille_code', 'maille_nom']).aggregate(aggregations).reset_index()
```